dbms/src/Storages/DeltaMerge/File/DMFileReader.cpp

```diff
@@ -142,147 +131,155 @@ bool DMFileReader::getSkippedRows(size_t & skip_rows)
-    auto & use_packs = pack_filter.getUsePacks();
+    // move forward next_pack_id and next_row_offset
+    size_t read_rows = getReadRows();
+    if (read_rows == 0)
+        return 0;
+
+    addSkippedRows(read_rows);
+    // When we read dmfile, if the previous pack is not read,
+    // then we should seek to the right offset of dmfile.
+    std::fill(must_seek.begin(), must_seek.end(), true);
+
+    scan_context->late_materialization_skip_rows += read_rows;
+    return read_rows;
+}
+
+// Get the number of rows to read in the next block
+// Move forward next_pack_id and next_row_offset
+size_t DMFileReader::getReadRows()
+{
+    const auto & use_packs = pack_filter.getUsePacksConst();
     size_t start_pack_id = next_pack_id;
+    // When read_one_pack_every_time is true, we can just read one pack every time.
+    // std::numeric_limits<size_t>::max() means no limit
+    size_t read_pack_limit = read_one_pack_every_time ? 1 : std::numeric_limits<size_t>::max();
     const auto & pack_stats = dmfile->getPackStats();
     size_t read_rows = 0;
     for (; next_pack_id < use_packs.size() && use_packs[next_pack_id] && read_rows < rows_threshold_per_read;
          ++next_pack_id)
     {
-        if (read_pack_limit != 0 && next_pack_id - start_pack_id >= read_pack_limit)
-            break;
-        if (enable_handle_clean_read && handle_res[next_pack_id] != expected_handle_res)
+        if (next_pack_id - start_pack_id >= read_pack_limit)
             break;
-
         read_rows += pack_stats[next_pack_id].rows;
     }
 
-    addSkippedRows(read_rows);
     next_row_offset += read_rows;
-
-    // When we read dmfile, if the previous pack is not read,
-    // then we should seek to the right offset of dmfile.
-    // So if skip some packs successfully,
-    // then we set the last pack to false to indicate that we should seek before read.
-    if (likely(read_rows > 0))
-        use_packs[next_pack_id - 1] = false;
-
-    scan_context->late_materialization_skip_rows += read_rows;
     return read_rows;
 }
 
 Block DMFileReader::readWithFilter(const IColumn::Filter & filter)
 {
+    /// 1. Skip filtered out packs.
+
     size_t skip_rows;
     if (!getSkippedRows(skip_rows))
         return {};
 
+    /// 2. Mark use_packs[i] = false if all rows in the i-th pack are filtered out by filter.
+
     const auto & pack_stats = dmfile->getPackStats();
     auto & use_packs = pack_filter.getUsePacks();
 
     size_t start_row_offset = next_row_offset;
-    size_t next_pack_id_cp = next_pack_id;
-
+    size_t start_pack_id = next_pack_id;
+    size_t read_rows = getReadRows();
+    RUNTIME_CHECK(read_rows == filter.size(), read_rows, filter.size());
+    size_t last_pack_id = next_pack_id;
     {
-        // Use std::find to find the first 1 in the filter, these rows before the first 1 should be skipped.
-        // For example, filter is [0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0, 0], each pack has 3 rows
```

Comment:
Why this example removed? Do we have a change in behaviour?

Reply:
I find `countBytesInFilter` is faster than `std::find`, so use `countBytesInFilter` instead which is faster and more simple.